packages/@uppy/dashboard/src/components/FileCard.js

```diff
@@ -7,57 +7,73 @@ class FileCard extends Component {
   constructor (props) {
     super(props)
 
-    this.meta = {}
+    this.state = {}
 
-    this.tempStoreMetaOrSubmit = this.tempStoreMetaOrSubmit.bind(this)
+    this.tempStoreMeta = this.tempStoreMeta.bind(this)
+    this.saveOnEnter = this.saveOnEnter.bind(this)
     this.renderMetaFields = this.renderMetaFields.bind(this)
     this.handleSave = this.handleSave.bind(this)
     this.handleCancel = this.handleCancel.bind(this)
   }
 
-  tempStoreMetaOrSubmit (ev) {
+  componentDidMount () {
     const file = this.props.files[this.props.fileCardFor]
+    const metaFields = this.props.metaFields || []
+
+    const storedMetaData = {}
+    metaFields.forEach((field) => {
+      storedMetaData[field.id] = file.meta[field.id]
+    })
+    this.setState(storedMetaData)
+  }
 
+  saveOnEnter (ev) {
     if (ev.keyCode === 13) {
+      const file = this.props.files[this.props.fileCardFor]
       ev.stopPropagation()
       ev.preventDefault()
-      this.props.saveFileCard(this.meta, file.id)
-      return
+      this.props.saveFileCard(this.state, file.id)
     }
+  }
 
+  tempStoreMeta (ev) {
     const value = ev.target.value
     const name = ev.target.dataset.name
-    this.meta[name] = value
+    this.setState({
+      [name]: value
+    })
+  }
+
+  handleSave (ev) {
+    const fileID = this.props.fileCardFor
+    this.props.saveFileCard(this.state, fileID)
+  }
+
+  handleCancel (ev) {
+    this.setState({})
```

Comment:
note this is a noop, it doesn't actually clear state

Reply:
Oh, right, only works because we're unmounting the component.
@arturi, maybe let's create the `this.state = { formState: {...} }` variable instead, and populate it with values from props right in the constructor?

Reply:
If we're guaranteed to be unmounting immediately after this, we could just not clear state and it will be gone when preact drops the component.

But the `formState` idea is the better option either way I think!

Reply:
Agreed, I think we shouldn't be clearing the state at all in `handleCancel()`, because we rely on calling the `constructor`/`componentDidMount` for assigning new variables anyway, clearing `formState` makes it seem like it matters.
I'd both remove the line 53 altogether, and scope form state under the `this.state.formState`.

Reply:
> Agreed, I think we shouldn't be clearing the state at all in `handleCancel()`, because we rely on calling the `constructor`/`componentDidMount` for assigning new state anyway,

My reasoning was that we are now, but at some point FileCard might be rendered but hidden, in that case it would make sense to clear on `cancel`. Maybe designing this for hypothetical cases isn’t a great idea ;) Changed, thanks!